File: dotnet/Base/GameX.Base/Store_Ubisoft.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;

namespace GameX.Store
{
    /// <summary>
    /// Store_Ubisoft
    /// </summary>
    internal static class Store_Ubisoft
    {
        internal static Dictionary<string, string> UbisoftPaths = new Dictionary<string, string>();

        static Store_Ubisoft()
        {
            // get dbPath
            var root = GetPath();
            if (root == null) return;
            var dbPath = Path.Combine(root, "settings.yaml");
            if (!File.Exists(dbPath)) return;
        }

        static string GetPath()
        {
            IEnumerable<string> paths;
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
            {
                // windows paths
                var home = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
                var search = new[] { "Ubisoft Game Launcher" };
                paths = search.Select(path => Path.Join(home, path));
            }
            else if (RuntimeInformation.OSDescription.StartsWith("android-")) return null;
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            {
                // linux paths
                var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
                var search = new[] { "??" };
                paths = search.Select(path => Path.Join(home, path));
            }
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
            {
                // mac paths
                var home = "/Users/Shared";
                var search = new[] { "??" };
                paths = search.Select(path => Path.Join(home, path));

            }
            else throw new PlatformNotSupportedException();
            return paths.FirstOrDefault(Directory.Exists);
        }
    }
}

File: Python/gamex/store_ubisoft.py
import os, platform, json

def getPath():
    system = platform.system()
    if system == 'Windows':
        # windows paths
        home = os.getenv('LOCALAPPDATA')
        search = ['Ubisoft Game Launcher']
        paths = [os.path.join(home, path) for path in search]
    elif system == 'Linux':
        # linux paths
        home = os.path.expanduser('~')
        search = ['??']
        paths = [os.path.join(home, path) for path in search]
    elif system == 'Darwin':
        # mac paths
        home = '/Users/Shared'
        search = ['??']
        paths = [os.path.join(home, path) for path in search]
    else: raise Exception(f'Unknown platform: {system}')
    return next(iter(x for x in paths if os.path.isdir(x)), None)
    
# get ubisoftPaths
ubisoftPaths = {}
root = getPath()
if root and os.path.exists(dbPath := os.path.join(root, 'settings.yaml')): 
    with open(dbPath, 'r') as f: body = f.read()
    gamePath = body[body.index('game_installation_path:') + 23:body.index('installer_cache_path')].strip()

    # query games
    for s in [s for s in os.listdir(gamePath)]:
        # print(s)
        pass
        
    #     with open(os.path.join(dbPath, s), 'r') as f:
    #         # add appPath if exists
    #         appPath = json.loads(f.read())['InstallLocation']
    #         if os.path.isdir(appPath): ubisoftPaths[s[:-5]] = appPath

# print(ubisoftPaths)
